pkg/controllers/disruption/consolidation.go

```diff
@@ -198,6 +204,69 @@ func (c *consolidation) computeConsolidation(ctx context.Context, candidates ...
 	}, nil
 }
 
+// Compute command to execute spot-to-spot consolidation if:
+//  1. The SpotToSpotConsolidation feature flag is set to true.
+//  2. For single-node consolidation:
+//     a. There are at least 15 cheapest instance type replacement options to consolidate.
+//     b. The current candidate is NOT part of the first 15 cheapest instance types inorder to avoid repeated consolidation.
+func (c *consolidation) computeSpotToSpotConsolidation(ctx context.Context, candidates []*Candidate, results *pscheduling.Results,
+	candidatePrice float64) (Command, error) {
+
+	// Spot consolidation is turned off.
+	if !options.FromContext(ctx).FeatureGates.SpotToSpotConsolidation {
+		if len(candidates) == 1 {
+			c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, "SpotToSpotConsolidation is disabled, can't replace a spot node with a spot node")...)
+		}
+		return Command{}, nil
+	}
+
+	// Since we are sure that the replacement nodeclaim considered for the spot candidates are spot, we will enforce it through the requirements.
+	results.NewNodeClaims[0].Requirements.Add(scheduling.NewRequirement(v1beta1.CapacityTypeLabelKey, v1.NodeSelectorOpIn, v1beta1.CapacityTypeSpot))
+	// All possible replacements for the current candidate compatible with spot offerings
+	instanceTypeOptionsWithSpotOfferings :=
+		results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions.Compatible(results.NewNodeClaims[0].Requirements)
+
+	// Possible replacements that are lower priced than the current candidate
+	results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions = filterByPrice(instanceTypeOptionsWithSpotOfferings, results.NewNodeClaims[0].Requirements, candidatePrice)
+
+	if len(results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions) == 0 {
+		if len(candidates) == 1 {
+			c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, "Can't replace spot node with a cheaper spot node")...)
+		}
+		logging.FromContext(ctx).With("nodeclaim", candidates[0].NodeClaim.Name).Debugf("can't replace spot node with a cheaper spot node.")
```

Comment:
Should we debug log here? This seems like this is going to be extremely noisy as it is evaluating NodeClaims for consolidation. Seems like the event should probably be enough here.

Reply:
If noisy, we should be cautious about the perf impact of spammy events, too (unless they appropriately deduped).

Reply:
> If noisy, we should be cautious about the perf impact of spammy events, too

It's a good point. Unconsolidatable events are deduped on a 15m timeout based on their UID, so realistically regardless of the reason we shouldn't be cycling a lot of events here.